File: app/lib/llm/custom-models.ts
// 模块说明：定义用户自定义模型（聊天 + 媒体）及其 SQLite API 数据结构。
import type { LlmProviderId } from "./types";

export type MediaMode = "text-to-image" | "text-to-video" | "image-edit";
export type MediaProtocol =
  | ""
  | "qwen-image-sync"
  | "volcengine-image"
  | "volcengine-video-async";
export type MediaOutputKind = "" | "image" | "video";

export interface CustomModelInput {
  name: string;
  provider: LlmProviderId;
  /** 原样发送给供应商 API 的 model 值。 */
  model: string;
  /** 可填写 Base URL 或完整接口地址。 */
  baseUrl?: string;
  includeInAuto: boolean;
  autoPriority: number;
  supportsVision: boolean;
  /** 声明为媒体模型时填写；空数组表示纯聊天模型。 */
  mediaModes: MediaMode[];
  mediaProtocol: MediaProtocol;
  mediaOutputKind: MediaOutputKind;
}

export interface CustomModelRecord
  extends Omit<CustomModelInput, "baseUrl"> {
  id: string;
  /** SQLite 中允许为空，服务端 JSON 会返回 null。 */
  baseUrl: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface CustomModelListResponse {
  models: CustomModelRecord[];
}

export interface CustomModelMutationResponse {
  model: CustomModelRecord;
}

File: backend/schemas/custom_models.py
"""用户自定义模型接口的数据结构（聊天 + 媒体）。"""

from __future__ import annotations

from typing import Literal

from pydantic import Field, field_validator

from backend.schemas.common import FlexibleModel

ProviderId = Literal[
    "qwen",
    "openai",
    "gemini",
    "deepseek",
    "glm",
    "kimi",
    "doubao",
]

MediaMode = Literal["text-to-image", "text-to-video", "image-edit"]
MediaProtocol = Literal["", "qwen-image-sync", "volcengine-image", "volcengine-video-async"]
MediaOutputKind = Literal["", "image", "video"]


class CustomModelInput(FlexibleModel):
    """新增或修改自定义模型时使用的请求体。"""

    name: str = Field(min_length=1, max_length=80)
    provider: ProviderId
    model: str = Field(min_length=1, max_length=160)
    base_url: str | None = Field(default=None, alias="baseUrl", max_length=500)
    include_in_auto: bool = Field(default=True, alias="includeInAuto")
    auto_priority: int = Field(default=10, alias="autoPriority", ge=1, le=9999)
    supports_vision: bool = Field(default=False, alias="supportsVision")
    # 媒体能力：空数组表示纯聊天模型；填了 modes 的模型会出现在媒体选择器里。
    media_modes: list[MediaMode] = Field(default_factory=list, alias="mediaModes")
    media_protocol: MediaProtocol = Field(default="", alias="mediaProtocol")
    media_output_kind: MediaOutputKind = Field(default="", alias="mediaOutputKind")

    @field_validator("name", "model")
    @classmethod
    def strip_required_text(cls, value: str) -> str:
        """去除首尾空白，并拒绝只有空格的名称或模型值。"""

        normalized = value.strip()
        if not normalized:
            raise ValueError("字段不能为空")
        return normalized

    @field_validator("base_url")
    @classmethod
    def validate_base_url(cls, value: str | None) -> str | None:
        """只允许保存 HTTP(S) Base URL，避免协议拼写错误。"""

        normalized = (value or "").strip().rstrip("/")
        if not normalized:
            return None
        if not normalized.startswith(("https://", "http://")):
            raise ValueError("Base URL 必须以 http:// 或 https:// 开头")
        return normalized

    @field_validator("media_modes")
    @classmethod
    def validate_media_modes(cls, value: list[MediaMode]) -> list[MediaMode]:
        """媒体协议/输出类型只在声明了媒体模式时才有意义，否则清空。"""

        if not value:
            return []
        return list(dict.fromkeys(value))


class CustomModelRecord(CustomModelInput):
    """返回给前端的自定义模型记录。"""

    id: str
    created_at: str = Field(alias="createdAt")
    updated_at: str = Field(alias="updatedAt")
